system-parachains/asset-hubs/asset-hub-polkadot/src/xcm_config.rs

```diff
@@ -631,18 +628,60 @@ pub mod bridging {
 				UniversalAliases::get().contains(alias)
 			}
 		}
-
-		/// Reserve locations filter for `xcm_executor::Config::IsReserve`.
-		/// Locations from which the runtime accepts reserved assets.
-		pub type IsTrustedBridgedReserveLocationForConcreteAsset =
-			matching::IsTrustedBridgedReserveLocationForConcreteAsset<
-				UniversalLocation,
-				(
-					// allow receive KSM from AssetHubKusama
-					xcm_builder::Case<KsmFromAssetHubKusama>,
-					// and nothing else
-				),
-			>;
+		/// Allow any asset native to the Kusama ecosystem if it comes from Kusama Asset Hub.
+		pub type KusamaAssetFromAssetHubKusama =
+			RemoteAssetFromLocation<StartsWith<KusamaEcosystem>, AssetHubKusama>;
+
+		// TODO: get this from `assets_common v0.17.1` when SDK deps are upgraded
+		/// Accept an asset if it is native to `AssetsAllowedNetworks` and it is coming from
+		/// `OriginLocation`.
+		pub struct RemoteAssetFromLocation<AssetsAllowedNetworks, OriginLocation>(
+			sp_std::marker::PhantomData<(AssetsAllowedNetworks, OriginLocation)>,
+		);
+		impl<
+				L: TryInto<Location> + Clone,
+				AssetsAllowedNetworks: Contains<Location>,
+				OriginLocation: Get<Location>,
+			> ContainsPair<L, L> for RemoteAssetFromLocation<AssetsAllowedNetworks, OriginLocation>
+		{
+			fn contains(asset: &L, origin: &L) -> bool {
+				let latest_asset: Location = if let Ok(location) = asset.clone().try_into() {
+					location
+				} else {
+					return false;
+				};
+				let latest_origin: Location = if let Ok(location) = origin.clone().try_into() {
+					location
+				} else {
+					return false;
+				};
+				let expected_origin = OriginLocation::get();
+				// ensure `origin` is expected `OriginLocation`
+				if !expected_origin.eq(&latest_origin) {
+					log::trace!(
+						target: "xcm::contains",
+						"RemoteAssetFromLocation asset: {:?}, origin: {:?} is not from expected {:?}",
+						latest_asset, latest_origin, expected_origin,
+					);
+					return false;
+				} else {
+					log::trace!(
+						target: "xcm::contains",
+						"RemoteAssetFromLocation asset: {latest_asset:?}, origin: {latest_origin:?}",
+					);
+				}
+
+				// ensure `asset` is from remote consensus listed in `AssetsAllowedNetworks`
+				AssetsAllowedNetworks::contains(&latest_asset)
```

Comment:
You can use the conditional `let` syntax to define the resolved `asset` and `origin` variables. Much less verbosy (and quite more readable, IMHO).

```suggestion
				let Ok(asset) = asset.clone().try_into() else {
					return false;
				};
				let Ok(origin) = origin.clone().try_into() else {
					return false;
				};

				let expected_origin = OriginLocation::get();
				// ensure `origin` is expected `OriginLocation`
				if !expected_origin.eq(&origin) {
					log::trace!(
						target: "xcm::contains",
						"RemoteAssetFromLocation asset: {asset:?}, origin: {origin:?} is not from expected {expected_origin:?}"
					);
					return false;
				} else {
					log::trace!(
						target: "xcm::contains",
						"RemoteAssetFromLocation asset: {asset:?}, origin: {origin:?}",
					);
				}

				// ensure `asset` is from remote consensus listed in `AssetsAllowedNetworks`
				AssetsAllowedNetworks::contains(&asset)
```

> Note 1: Consider if it's possible to externalize this helper struct to a common crate (to avoid duplication and help to maintainability of this codebase).
> Note 2: Kudos for adding `trace` logs. This is ALWAYS appreciated.